proposals/module-types/Explainer.md

```diff
@@ -0,0 +1,745 @@
+`.wasm` module for portable distribution.
+
+Additionally, on the Web and other hosts that provide runtime instantiation
+APIs, nested modules allow an outer module to bundle module code that is to be
+instantiated dynamically by passing a first-class module reference (created via
+`ref.module`) to the runtime instantiation API.
+
+
+### Module and Instance Exports
+
+Lastly, symmetric to all other kinds of definitions, modules and instances can
+be exported. For example:
+```wasm
+(module
+  (import "a" (module $a ...))
+  (module $b ...)
+  (import "c" (instance $c ...))
+  (instance $d ...)
+
+  (export "e1" (module $a))
+  (export "e2" (module $b))
+  (export "e3" (instance $c))
+  (export "e4" (instance $d))
+)
+```
+Therefore, module and instance types can appear in both the imports and exports
+of module types and instance types.
+
+
+### Summary
+
+To summarize the proposed changes (all changes in both text and binary format):
+* The `module` field of [`import`] become optional (allowing single-level
+  imports). (How to encode this in the [import section] is an interesting
+  question.)
+* New `module` and `instance` type constructors are added that may be used to
+  define types in the [type section].
+* A new `alias` case is added to type definitions allowing the import of parent
+  modules' type definitions.
+* New `module` and `instance` cases are added to [`importdesc`], with the
+  type indicated by `typeidx`, like `func`.
+* A new `module` section is added which contains a sequence of either module
+  definitions or `alias` definitions (of parents' modules). In terms of the
+  binary format, a nested module definition has the same binary format as a
+  non-nested `.wasm` module. Thus, the [`module`] binary format production
+  becomes recursive.
+* A new `instance` section is added which contains a sequence of instance
+  definitions. Like [`global`] definitions, an `instance` definition contains
+  an `instance` type and an `init` expression which is validated to match.
+* A new `instance.instantiate` instruction is added which is only allowed in
+  `instance` `init` expressions.
+* New `ref.instance`, `ref.module`, `ref.memory`, `ref.table` and `ref.global`
+  instructions are added to the general instruction set.
+* New module, instance, memory, table and global reference types are added to
+  [`reftype`].
+* New `module` and `instance` cases are added to [`exportdesc`].
+
+
+## Use Cases Revisited
+
+Worked examples of the [above use cases](#use-cases) are given in separate docs:
+* [Link-Time Virtualization](Example-LinkTimeVirtualization.md)
+* [Shared-Everything Dynamic Linking](Example-SharedEverythingDynamicLinking.md)
+
+## Additional Requirements Revisited
+
+Reconsidering the requirements stated [above](#additional-requirements):
+* **No GC dependency**: The lifetimes of nested instances are necessarily the
+  same as the lifetime of their parent instance. Thus, this proposal introduces
```

Comment:
Is that actually true? If they are exported, can't they outlive their parent?

Reply:
Good point; I'll nuance this.

Reply:
Actually, the latter sentences mostly address this by adding caveats for "if the host...", but I can still tweak it a bit.

Reply:
Fair enough, but instead of "same" shouldn't it say "as least as long" then?

Reply:
Yup, my "tweak" ended up rewriting the para and indeed "same" is gone.